third_party/xla_client/metrics_analysis.cc

```diff
@@ -0,0 +1,204 @@
+    // XRT GetMetrics call is relatively expensive.
+    if (++counter_ % run_every_n_ != 0) {
+      return {Analysis::Symptom::kNormal};
+    }
+    counter_ = 0;
+    CounterData* step = GetCounter("MarkStep");
+    if (!step) {
+      return {Analysis::Symptom::kNormal};
+    }
+
+    std::stringstream ss;
+    xla::int64 step_count = step->Value();
+    auto xrt_metrics = ComputationClient::Get()->GetMetrics();
+    for (auto const& kv : metric_name_thresholds_) {
+      auto it = xrt_metrics.find(kv.first);
+      if (it == xrt_metrics.end()) {
+        continue;
+      }
+      xla::Metric xrt_metric = it->second;
+      std::ldiv_t res;
+      if (xrt_metric.int64_value) {
+        int64 metric_count = *xrt_metric.int64_value;
+        res = std::div(metric_count, step_count);
+      } else if (xrt_metric.percentile) {
+        size_t metric_count = (*xrt_metric.percentile).total_samples;
+        res = std::div(metric_count, step_count);
+      } else {
+        continue;
+      }
+      if (DivToFloat(res, step_count) > kv.second) {
+        ss << kv.first << " (" << DivToFloat(res, step_count) * step_count
+           << " counts), ";
+      }
+    }
+
+    std::string repr = ss.str();
+    if (!repr.empty()) {
+      return {
+          Analysis::Symptom::kMetricTooFrequent,
+          absl::StrFormat("Following metrics too frequent: %sduring %zu steps. "
+                          "Note: XRT metrics follow the lifecycle of the TPU "
+                          "so you may need "
+                          "to restart the TPU for fresh metrics.",
+                          repr, step_count),
+      };
+    }
+    return {Analysis::Symptom::kNormal};
+  }
+
+ private:
+  std::map<std::string, float> metric_name_thresholds_;
+  bool is_xrt_metric_;
+  int run_every_n_;
+  int counter_;
+};
+
+class UnloweredOp : public Analyzer {
+ public:
+  Analysis Run() override {
+    std::stringstream ss;
+    MetricsArena* arena = MetricsArena::Get();
+    arena->ForEachCounter([&ss](const std::string& name, CounterData* data) {
+      if (absl::StrContains(name, "aten::") &&
+          name != "aten::_local_scalar_dense") {
+        ss << name << ", ";
+      }
+    });
+
+    std::string repr = ss.str();
+    if (!repr.empty()) {
+      return {Analysis::Symptom::kUnloweredOp,
+              absl::StrCat("Op(s) not lowered: ", repr,
+                           " Please open a GitHub issue with the above op "
+                           "lowering requests.")};
+    }
+    return {Analysis::Symptom::kNormal, repr};
+  }
+};
+
+std::vector<Analyzer*>* GetAnalyzers() {
+  static std::vector<Analyzer*>* analyzers = new std::vector<Analyzer*>{
+      new MetricFrequency("CompileTime", 0.5f, 10),
+      new MetricFrequency("TransferFromServerTime", 0.5f),
+      new MetricTime("CompileTime", 300e9),
+      new MetricTime("ExecuteTime", 30e9),
+      new UnloweredOp(),
+      new XrtMetricFrequency({{"XrtTryFreeMemory", 0.1f},
```

Comment:
maybe in the future we can make it configurable by env var.

Reply:
Was hoping to avoid having too many/complex env vars, but depending on feedback let's do that.